nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-client-services/src/main/java/org/apache/nifi/kudu/KuduControllerService.java

```diff
@@ -0,0 +1,344 @@
+                .defaultOperationTimeoutMs(Integer.parseInt(operationTimeout))
+                .defaultAdminOperationTimeoutMs(Integer.parseInt(adminOperationTimeout))
+                .build();
+    }
+
+    public void flushKuduSession(final KuduSession kuduSession, boolean close, final List<RowError> rowErrors) throws KuduException {
+        final List<OperationResponse> responses = close ? kuduSession.close() : kuduSession.flush();
+
+        if (kuduSession.getFlushMode() == SessionConfiguration.FlushMode.AUTO_FLUSH_BACKGROUND) {
+            rowErrors.addAll(Arrays.asList(kuduSession.getPendingErrors().getRowErrors()));
+        } else {
+            responses.stream()
+                    .filter(OperationResponse::hasRowError)
+                    .map(OperationResponse::getRowError)
+                    .forEach(rowErrors::add);
+        }
+    }
+
+    protected KerberosUser loginKerberosUser(final String principal, final String keytab) throws LoginException {
+        final KerberosUser kerberosUser = new KerberosKeytabUser(principal, keytab);
+        kerberosUser.login();
+        return kerberosUser;
+    }
+
+    public final void closeClient() throws Exception {
+        try {
+            if (this.asyncKuduClient != null) {
+                getLogger().debug("Closing KuduClient");
+                this.asyncKuduClient .close();
+                this.asyncKuduClient  = null;
+            }
+        } finally {
+            if (kerberosUser != null) {
+                kerberosUser.logout();
+                kerberosUser = null;
+            }
+        }
+    }
+
+    @OnDisabled
+    public void shutdown() throws Exception {
+       this.closeClient();
+    }
+
+    public void scan(ProcessContext context, ProcessSession session, KuduTable kuduTable, String predicates, List<String> projectedColumnNames, ResultHandler handler) throws Exception {
+        AsyncKuduScanner.AsyncKuduScannerBuilder scannerBuilder = this.asyncKuduClient.newScannerBuilder(kuduTable);
+        final String[] arrayPredicates = (predicates == null || predicates.isEmpty() ? new String[0] : predicates.split(","));
+
+        for(String column : arrayPredicates){
+            if (column.contains(":")) {
+                final String[] parts = column.split(":");
+                addPredicate(scannerBuilder, kuduTable, parts[0], parts[1]);
+            }
+        }
+
+        if(!projectedColumnNames.isEmpty()){
+            scannerBuilder.setProjectedColumnNames(projectedColumnNames);
+        }
+
+        AsyncKuduScanner scanner = scannerBuilder.build();
+        while (scanner.hasMoreRows()) {
+            handler.handle(scanner.nextRows().join());
+        }
+    }
+
+    @VisibleForTesting
+    public void buildPartialRow(Schema schema, PartialRow row, Record record, List<String> fieldNames) {
+        for (String colName : fieldNames) {
+            int colIdx = this.getColumnIndex(schema, colName);
+            if (colIdx != -1) {
```

Comment:
What happens to missing columns?

Reply:
it will be Ignored. it only takes fields which are in kudu table